sktime/forecasting/ensemble_algorithms.py

```diff
@@ -0,0 +1,234 @@
+        time_length = expert_predictions.shape[1]
+
+        for i in range(time_length):
+            loss_vector = np.array([
+                          self.loss_func([prediction], [actual_values[i]])
+                          for prediction in expert_predictions[:, i]
+                          ])
+
+            average_loss = np.dot(self.weights, loss_vector)
+
+            instant_regret = (average_loss - loss_vector)
+            self.R += instant_regret
+            self._update_weights(low_c=low_c)
+
+    def _update_weights(self, low_c=0.01):
+        """Updates the weights on each of the experts by performing a potential
+        function update with a root-finding search. low_c represents the lower
+        bound on the window that the root finding is occuring over.
+
+        Parameters
+        ----------
+        low_c : float
+        """
+
+        # Calculating Normalizing Constant
+        R_plus = np.array(list(map(lambda x: 0 if 0 > x else x, self.R)))
+        normalizing_R = np.max(R_plus)
+
+        R_plus /= normalizing_R
+
+        low_c = low_c
+        high_c = (max(R_plus)**2)/2
+
+        def pot(c):
+            """Internal Potential Function
+
+            Parameters
+            ----------
+            low_c : float
+
+            Returns
+            -------
+            potential: float
+            """
+            return np.mean(np.exp((R_plus**2)/(2*c)))-np.e
+
+        c_t = bisection(low_c, high_c, pot)
+
+        def prob(r, c_t):
+            """Internal Probability Function
+
+            Parameters
+            ----------
+            r : float
+            c_t : float
+
+            Returns
+            -------
+            prob : float
+            """
+            return (r/c_t)*np.exp((r**2)/(2*c_t))
+
+        self.weights = np.array([prob(r, c_t) for r in R_plus])
+        self.weights /= np.sum(self.weights)
+
+
+class NNLSEnsemble(EnsembleAlgorithm):
+    """ Wrapper class to perform a non-negative least squares to fit to the experts
+    Keeps tracks of all observations seen so far and fits to it
+
+    Parameters
+    ----------
+    n : int, number of experts
+    loss_func : function, loss function
+    """
+
+    def __init__(self, n=10, loss_func=None):
+        super().__init__(n=n, loss_func=loss_func)
+        self.total_expert_predictions = np.empty((n, 0))
+        self.total_actual_values = np.empty((0))
+
+    def _update(self, expert_predictions, actual_values):
+        self.total_expert_predictions = np.concatenate(
+            (self.total_expert_predictions, expert_predictions), axis=1)
+        self.total_actual_values = np.concatenate(
+            (self.total_actual_values, actual_values))
+        weights, loss = nnls(self.total_expert_predictions.T,
+                             self.total_actual_values)
+        self.weights = weights
+
+
+"""
+Helper Functions
+"""
+
+
+def bisection(low, high, function, threshold=1e-8):
```

Comment:
instead of writing your own bisection method, why not use the well-curated and highly optimized `scipy.optimize.bisect`?

Furthermore, the potential function that you are applying it to below is smooth, so you could use a method with a derivative, such as Newton-Raphson, or the Brent method - which `scipy.optimize` claims in docs is "the best method" for hte problem.

Reply:
I think I went with bisection because I know that it was robust and could solve the problem relatively quickly. I can try to implement other iterations that use other root finding methods in `scipy.optimize`, but am very comfortable with how bisection should perform.